### ISSUE TITLE
Handling the immutability

### ISSUE THREAD

**Issue by christianalfoni** [arqex/freezer#3]
Hi @arqex !

Great work on this lib! Since it pretty much uses the same API signature as native it can be used in a lot of different places, even Angular, where I am doing a little experiment now.

I have this one issue though. I want my Freezer instance to be the holder of my application state. This actually does work really great, but doing updates gives some challenges and feels a bit confusing.

```javascript
var store = new Freezer({
  items: []
});

// The store instance is made available to the app
var storeInstance = store.get();

// Somebody does an unwanted change somewhere in the app
storeInstance.items.push('foo');

// No worries, it is not available to anyone, that is great!
storeInstance.items // []

// But now I really do want to update the store instance. I grab the latest version
storeInstance = store.get();

// Do the change
storeInstance.items.push('bar');

// Update the instance
storeInstance = store.get(); 

// But it shows both changes
storeInstance.items // ['foo', 'bar']

```

To me it is very hard to reason about this. It seems like the Freezer instance is kind of a global where the first mutation to each branch is available on the next `get()`. This means than any part of your code actually can mutate the store. 

To me it seems that the `get()` method should be moved, let me explain with an example:

```javascript
// Returns the instance, as if `get()` is already called
var store = new Freezer({
  items: []
});

// Should create a brand new wrapper store where the change lives
store.items.push('foo');

// So the current store is still the same
store.items // []

// To actually grab the changes from the store you use `get()` on the mutation you did
store = store.items.push('bar').get();
store.items // ['bar']

```
This gives some advantages:

1. It is easier to reason about that a mutation to a store does not work. Neither with `store.foo = 'bar'` or `store.items.push('foo')`

2. It is only the specific mutation that is returned `var items = store.items.push('foo')`. 

3. To update the store itself you have to grab it after your mutation is done and update the instance `store = store.items.push('foo').get()`

This also allows you to completely lock the store inside a module and expose specific methods to do the mutations.

```javascript
var storeModule = (function () {
  var store = new Freezer({
    list: []
  });

  return {
    addItem: function (item) {
      store = store.items.push(item).get();
    },
    getItems: function () {
      return store.items;
    }
  };
}());

storeModule.addItem('foo');
storeModule.getItems() // ['foo']

```

I hope this made sense :-)

**Comment by christianalfoni** [comment]
In case you read the "ops" (commented by mistake) I am commenting again to notify that the issue is updated :-)

**Comment by arqex** [comment]
Hi @christianalfoni 

Thanks for your comments. At the beginning it is difficult to get used to work with freezer stores, even for me. But I think they are pretty consisitent and I am getting more confortable every day.

As you said the store is one, and any operation in a branch will update the whole store, so the whole app can react to the change. If you want to make local changes without affecting the rest of the store, or if your framework does, the better you can do is to get the use the value of the branch instead the branch itself:

```js
var store = new Freezer({
  items: []
});

// The store instance is made available to the app
var storeInstance = store.get();

// I will have a copy of the store that I can modify
var localnstance = storeInstance.items.toJS();

locaInstance.push(1).push(2).push(3);

localInstance; // [1,2,3];
storeInstance; // {items:[]}
store.get(); // {items:[]}

// I like the changes, so i can reintegrate them into the store
// I will refresh the storeInstance with the new node
storeInstance = storeInstance.set('items', localInstance);

storeInstance; // {items: [1,2,3]}
store.get() === storeInstance; // true
```
So it is needed to keep in mind that the store is one, and only make changes when you really want to update the store. 

Said so, I would reduce as much as possible local copies of nodes. They, like local states, make more complex reasoning about the application. It is always easier to update the store completely, and having one source of truth. Modifying the whole store should be the default behaviour ( whithout using `get` ) and making changes locally the exception.

If your bother about rendering performance, all the updates made to the store in the same tick just trigger one 'update' event in the store, so your UI only will be refreshed once.

It is possible to work with store detached nodes, in a similar way you make in your examples, but I wouldn't recommend it at all:
```js
storeInstance = store.get().items;

// Let's update the store
updated = storeInstance.push(1);
storeInstance; // []
updated; // [1]

// storeInstance is now detached from the store
storeInsance === store.get().items; // false

// You can update it as much as you want it won't affect the store
updated = storeInstance.push(2).push(3);
storeInstance; // [] is immutable
updated; // [2,3] // generated from storeInstance
store.get().items; // [1] // remains the same because storeInstance wasn't attached to the store
```
Again **this is not recommended** because it is really easy to lose the track of what is attached and what not, and you will find yourself debugging a lot to find out why your app doesn't refresh.

It is not easy to make your mind to the one store paradigm, but it worths it.

**Comment by arqex** [comment]
I have to say that the best part of immutability, IMHO, it is that it forces you to change your way of thinking. 

Once you have changed it, it is not important if the data is mutable or not, you won't try to change it directly. I am saying so because freezing all the store have a big penalty on performance, and probably I will add the option of disable the freezing in the future. The store can be used exactly the same way without being immutable ( in fact, immutable.js stores are mutable ).

**Comment by christianalfoni** [comment]
HI @arqex and thanks for taking the time to give such a detailed answer :-)

I see your point with twisting your mindset a bit, so as I am working on that I hope it is okay I ask a couple of more questions :-)

I do not quite see how you get control of unwanted updates:

```javascript
var store = new Freezer({
  items: []
});

// To get any data out of my store I have to get it
var storeInstance = store.get();

// I pass my storeInstance around the application to grab state
var myItemsLabel = 'I have ' + storeInstance.items.length + ' items';

// Suddenly an unwanted change is made to the store, but the current
// state of the store is of course not affected
storeInstance.items.push('unwanted');

// But then some business logic that is supposed to update the store, does so
{
  addItem: function (item) {
    storeInstance.items.push(item);
    storeInstance = store.get();
  }
}

// Now I have two items in my items list, but I only want one
storeInstance.items; // ['unwanted', 'somethingElse']

// I am unable to retrieve the new version of the store based on a 
// nested updates. Shallow updates is no problem though. There I could do:
var store = new Freezer({
  foo: 'bar'
});

var storeInstance = store.get();
storeInstance = store.set('foo', 'bar2');

// The only way to change the state of the store is to replace it. I would never use `store.get()`
// again, after the initial setup. But I can not do something like this on a nested update. 
// And that is what I need... or at least think I need :-)

```
Am I just thinking about this the wrong way? I just want all my updates to work just like the last example, even nested updates. If I always have to use `store.get()` to retrieve the new version of a store any part of my application can do updates.

**Comment by arqex** [comment]
I think I don't fully get what you mean. Do you need a way of replacing just a node by another?
If you mean that, I am afraid that freezer doesn't provide that method, but it shouldn't be difficult to implement. I have found myself requiring a method like that.

What I don't understand is why a part of your application would make an unwanted change. If it is some component that you don't fully control ( a vendor component ) the best thing is to create a copy of the node, with `toJS()` as I recommended, and give it to the component.

With freezer any component is able to update its part of the store, it is like they have its own domain and they are responsible of it.

Is that what you meant?

**Comment by christianalfoni** [comment]
Hi again @arqex ,

Thanks for following up on this, I am not explaining my self too well here :-) The "unwanted" change was bad naming on my part.

I think where we see things differently is that you want components to be able to update the store directly, but I want all changes to go through actions, in a more traditional FLUX pattern.

But maybe you could give an example of how you would create a store and use it in a React component? Is it something like this?

```javascript
var store = require('./store.js');

var Mycomponent = React.createClass({
  getInitialState: function () {
    return {
      items: store.get().items
    };
  },
  componentWillMount: function () {
    store.getListener().on('update', this.updateState);
  },
  componentWillUnmount: function () {
    store.getListener().off('update', this.updateState);
  },
  updateState: function () {
    this.setState({
      items: store.get().items
    });
  },
  render: function () {
    return <h1>Hello world</h1>
  }
});
```

Then if I understand your intended use I can better explain what I want to achieve :-)

**Comment by arqex** [comment]
I am preparing a tutorial creating a JSON editor, here you have the code:

http://jsbin.com/lelurusuca/2/edit?html,js,output

In a simple example, if you use the store update methods in the components themselves, it seems like the component is the one than updates the store. But it is not like that. What you do using `push` is requesting the change. The update is completed by Freezer, starting from the top of the store and the properties are passed down again through the componoents.

The data flows in only one direction as Flux recommends, you can see Freezer as a store with some predefined actions and action callbacks.

In a more complex example you would need to have real actions. Then freezer nodes can be passed as the payload of the action callback, making easier the update of the store.

I think that freezer modifier methods shouldn't be used inside the components directly, but I am not a big fan of dispatcher and actions, that is why we were looking for communication ways in the 'reactive' discusson we had in the 'Curxor' forum.

I am also preparing an article about component communication, but I have not enough time talk about everything I would like :D

**Comment by christianalfoni** [comment]
Aha, I definitely see your point with the modifiers still honoring the one way flow. But as you say, initiating state change within a component is not ideal. One thing is just reasoning about your code. It is very nice to look at your action(s) file to get a grip on what state changes are actually possible in this application. Also making more complex state changes that includes async stuff.

I created this crazy angular experiment, using freezer, have an article on it. Can send it to you when it is published. It is kinda cool, though very experimental. Trying to convince Angular developers that they should at least not turn their back on React JS and FLUX :-)

Looking at React JS conf talks etc. there seems to still be a missing part and that is passing state from top component down through the app. It is too much work defining all the props, making sure it is passed down correctly. The new "withContext" turned out not to be a good way to do it either. 

I think overall the best solution would be to have a single immutable store, like freezer. Pass all the state down through the top component,  but with a different concept than props. Not sure how it should work, but then you would have a single input, single event and it would be supereasy to do isomorphism too. The article I wrote on Baobab points to a very effective way of handling state, though each component has a relationship to the store. It would be nice to remove that relationship completely.

So yeah, the reason I got so caught up in this is that I say in my article that "Freezer is completely immutable", but the way I use it, it becomes untrue. The reason is that every time an action is triggered the main store is updated with `.get()`. If a component had run a `push`, `set` etc. on any of the state from the store, those changes would be included when the next action runs.

Anyways, no biggie, just an experiment. Will be exciting to see where all this state handling lands. Facebooks Relay etc. :-)

Thanks for contributing!

**Comment by christianalfoni** [comment]
Just wanted to mentioned I a put a lib inspired by Freezer on github yesterday. Thanks for answering my qeustions etc. Learned a lot looking at Freezer code and trying to build my own thing :-)

https://github.com/christianalfoni/immutable-store

**Comment by arqex** [comment]
Ey @christianalfoni 

Your immutable-store is really interesing too. You return the full store on change, it is an interesting idea I had in mind when I started with curxor. But at the end I decided to return only the updated part in order to encourage encapsulation.

As for the relations of the components with the store, using this kind of immutable store you can eliminate it completely. Don't use the updating methods of the store inside the component, just dispatch an action with the store part ( the component don't even know that it is a store part ) and use it inside an action callback.

I have just written an article about component communication. In case you want to make your components independent from the actions too, I think it is the way to go :)

http://arqex.com/1028/better-flux-dom-events

Looking forward to reading your article too.

Cheers,
Javi

**Comment by christianalfoni** [comment]
Hi @arqex! :-)

Thanks for the feedback! Its fun to play around with this concept and see where it takes us.

Yeah, the good thing about Freezer is that you have this really simple and elegant native like API ... which I was a bit mind-blown and heavily inspired by ;-) ... and you do not have to override the current store reference, which you have to in immutable-store. You just have to make sure that whenever you need the latest version of the tree you do a "get()". 

The thing that I wanted to solve is that it should be completely impossible to mutate the store from a component. You have to specifically call an API to set the new version of the store. That way nobody can accidentally do a mutation that will suddenly pop up somewhere later. As you say, you should pass the part of the store that should be mutated through an action, but what if a developer suddenly does the mutation right in the component? It would pop up the next time a "get()" was run on a store. It is a bit paranoid, but I wanted to figure out a way to ignore mutations that are not specifically set :-)

About your article. Really great stuff, it makes you think :-) I think there are a couple of challenges though:

1. Node JS does not have document, so you can not run those components on the server side

2. It is not always a component listening to an event is a parent. It could be a sibling, which would not receive the event. This is a big issue with Angular where you have scoping. $emit up the tree, and $broadcast down. You very quickly get into an issue with siblings not getting events

3. Is not the `document` just as much an external dependency as `MyEventEmitter`? ;-)

That said, I still think this is a very interesting concept!

**Comment by arqex** [comment]
Hehehe, I see your paranoid point ;) Maybe having the tools to update the store in every node adds the possibility of unwanted changes, but at least all the store know about the changes and can update the views accordingly.

This event approach is about to communicate user interaction more than to render the views, that's why you hardly will need to use it on the server. In case you need it in the server things are difficult, since I wouldn't know what to use as a hub. In browsers I use `document` because it is a natural event hub and I don't thing see it as an external dependency. It is more a environment tool, the same way that you can use `Array` you can use `document` in browsers.

If we could use custom React.js synthetic events, this approach would be cross-environment.

Sibling dependency has always been a problem to me and I always try to avoid it. If necessary, it is better done through the parent to let the data keep flowing in the same direction.

**Comment by christianalfoni** [comment]
Yeah, I see your point with `document` being less of a dependency than a JS module. It would indeed be great if React JS had a concept of passing events up the component tree! Just to get rid of passing callbacks even.

Btw, I am twisting and turning in bed thinking about store injection... true story :-) I really really like the concept of having the one application state tree that you pass from the top component. It makes all the components really dumb and you can pre-process the application very easily on the server. The thing is, there is no good way to extract the state from this single injected tree.

Now you currently do something like this:

```javascript
React.render(<App store={store}/>, document.body);
``` 

But this of course requires you to pass that store as a prop down to all your components. It is incredibly verbose. I also mentioned this `withContext` method:

```javascript
React.withContext(store, function () {
  React.render(<App/>, document.body);
});
```
That allows any sub component to extract the state, even though their parent does not.  BUT, if you use the PureRenderMixin you get into problems.

I was thinking of events. What if you for example passed a Freezer store down as a context. All components would initially render and use the contents of the stores. The challenge is to get them to update. But Freezer has the possibility grab a listener from all the props. So what if:

```javascript
React.withContext(myFreezerStore.get(), function () {
  React.render(<App/>, document.body);
});

var App = React.createClass({
      contextTypes: {
        name: React.PropTypes.string.isRequired
    },
    componentWillMount: function () {
      this.context.name.getListener().on('update', this.update);
    },
   update: function () {
    this.setState({});
  },
  render: function () {
    return <h1>{this.context.name}</h1>
  }
});
```
Of course not a complete solution, but the thing is that this should work. By using a mixin it would be possible to automate this, also make it work on the server:

*app.js*
```javascript
var App = React.createClass({
    mixins: [ContextUpdateMixin],
    contextTypes: {
        name: React.PropTypes.string.isRequired
    },
  render: function () {
    return <h1>{this.context.name}</h1>
  }
});
```
The mixin goes through all properties on the "context" object and registers listeners. If the property does not have a listener (running on server) it ignores it.

*on server*
```javascript
React.withContext({name: 'christian'}, function () {
  React.render(<App/>, document.body);
});
```

*on client*
```javascript
React.withContext(myFreezerWithSameState.get(), function () {
  React.render(<App/>, document.body);
});
```
And now listeners are registered and components will individually update when changes occurs. Not passing a single prop or using "outside" dependencies, except the mixin of course.

Okay... just had to write this down, heh

**Comment by arqex** [comment]
That way you save yourself of passing the store down. But every component has access to the complete store, and they should known where its domain is inside the store. That might be tedious too.

**Comment by christianalfoni** [comment]
Hm yeah, of course, did not think that you can only point to property on top level of store. Hm hm. Ah well, let that churn a bit more first. Thanks for hearing me out :-)

**Comment by ArnoBuschmann** [comment]
@christianalfoni 
>
But this of course requires you to pass that store as a prop down to all your components. It is incredibly verbose.

While not having read this entire post, I guess, the problem you are talking about here relates to a solution which initially got called "reference cursors" in Om:

https://github.com/omcljs/om/wiki/Advanced-Tutorial#reference-cursors

>
reference cursors are... very natural and straightforward, the component did not need to take its data from a parent nor did it need to receive it via a channel.

This concept also got implemented in Immstruct:

https://github.com/omniscientjs/immstruct#references

**Comment by christianalfoni** [comment]
Hi guys,

Yup, reference cursors is awesome :-)

I wrote on article on this, published today: [True isomorphic apps with React and Baobab](http://christianalfoni.github.io/javascript/2015/03/01/true-isomorphic-apps-with-react-and-baobab.html). I was initially writing it for freezer, but I had some problems with listeners on values that got nullified and then got a new value. Can freezer handle that?

**Comment by arqex** [comment]
Ey nice article!

Good idea to pass down the actions using the context too! As for isomorphic apps there was a thread were Arno and me discuss about a Router, a key part of isomorphic apps imho:

https://github.com/arqex/fluxify/issues/2

Maybe it inspires you.

The problem you had with freezer, can you reproduce it in some jsbin? I don't have problems nullifying.

Cheers,
Javi

**Comment by christianalfoni** [comment]
HI @arqex,

Sorry that I missed out on using Freezer as an example on the article. If I had given it some more work I would probably have figured out why it did not work. I just have a lot more experience with Baobab. That said, there should not be any reason why Freezer should not work with that pattern :-)

I am not sure what exactly happened, but the suspicion was that a registered listener to a value with be invalidated if the value was removed. But if that is not the case it must have been something else. Sorry to not being able to give you more details on that! I am not done with Freezer, so maybe I will let you know if I get into the issue again :-)

I will read the thread as soon as I got a few minutes extra, thanks for the tip!
